docs/advanced/runtime/docs/atomics-memory-model.md

```diff
@@ -0,0 +1,128 @@
+release semantics, the meaning of atomicity, and so on.
+
+The actual implementation of these operations in Mono is described at he end.
+
+## Semantics
+
+### Atomicity in the CLI
+
+Any load or store that is smaller than or equal to `IntPtr.Size` shall be atomic, but does not imply a barrier of any
+kind. Operations on 64-bit quantities are only atomic on 64-bit systems.
+
+The source/destionation address of a load/store operation must be properly aligned for the data type for the above
+guarantees to hold.
+
+If a load or store to an address happens at the same time as another load or store to that address but of a different
+size, all bets are off and no atomicity is guaranteed.
+
+These rules apply to high-level languages like C# and F# as they target the CLI.
+
+### `volatile.` prefix opcode in CIL
+
+When the `volatile.` prefix opcode is used in CIL, it imposes acquire/release semantics on the next non-prefix opcode.
+For loads, it results in acquire semantics. For stores, it results in release semantics.
+
+This prefix opcode has no effect on atomicity beyond the standard rules of the CLI.
+
+### `volatile` keyword in C#
+
+The `volatile` keyword in C# compiles down to `Thread.VolatileRead` and `Thread.VolatileWrite` calls. Nothing more,
+nothing less. This means loads or stores to `volatile` fields have the same semantics as these methods.
+
+C#'s `volatile` cannot be applied to 64-bit quantities because these methods do not guarantee atomicity for 64-bit
+quantities on 32-bit systems, and the `Volatile` class did not exist when the `volatile` keyword was designed. Today,
+`volatile` on 64-bit quantities could conceivably be compiled down to `Volatile.Read` and `Volatile.Write` calls.
+
+### `Thread` class
+
+The `VolatileRead` and `VolatileWrite` methods perform loads and stores with acquire and release semantics,
+respectively. They guarantee absolutely nothing about atomicity beyond the standard rules of the CLI. In effect, this
+means that the 64-bit overloads of these methods are not atomic on 32-bit systems.
+
+There is a quirk in the Microsoft implementation where these methods actually use the `MemoryBarrier` method to insert
+a barrier. This is stronger than a simple acquire or release barrier. We do the same for compatibility.
+
+The `MemoryBarrier` method inserts a full sequential consistency barrier.
+
+### `Volatile` class
+
+The methods on the `Volatile` class are all atomic regardless of system bitness, and result in acquire and release
+barriers for loads and stores respectively.
+
+The 64-bit methods on this class are not atomic with respect to loads or stores made through other means than the
+methods on this class and the `Interlocked` class. This is because such 64-bit operations may need to be implemented
+with a lock on 32-bit systems.
+
+The MSDN documentation lies about the C# compiler emitting calls to this class's methods. It does not.
```

Comment:
BTW, it may actually be intentional in the sense that the (32-bit) `Volatile` methods are what the `volatile` keyword in C# is actually specified to match semantically. It's technically a bug that `Thread.VolatileRead`/`Thread.VolatileWrite` have stronger barriers than specified, and I don't think they want new code to rely on this.